File: app/src/test/java/com/bluzi/babymonitor/monitor/CryCalibrationTest.kt
package com.bluzi.babymonitor.monitor

import com.bluzi.babymonitor.data.Settings
import com.bluzi.babymonitor.dsp.SR
import com.bluzi.babymonitor.dsp.analyzeWindow
import com.bluzi.babymonitor.dsp.babyCry
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

// ALRM-2 (one sensitivity dial) and ALRM-15/16/17 (learning from the parent's answers).
// The mapping and the learning are pure logic; detector integration runs the REAL analysis.

private const val WINDOW = 2048
private const val WINDOW_MS = WINDOW * 1000L / SR

/** Feed [seconds] of a real synthesised cry to a detector and report whether it ever alarmed. */
private fun cryFires(thresholdDb: Double, levelDb: Double, seconds: Double = 4.0): Boolean {
    val detector = BabyNoiseDetector(enabled = true, thresholdDb = thresholdDb)
    var now = 0L
    var fired = false
    val windows = (seconds * SR / WINDOW).toInt()
    val pcm = babyCry(WINDOW * windows)
    for (w in 0 until windows) {
        val chunk = pcm.copyOfRange(w * WINDOW, (w + 1) * WINDOW)
        now += WINDOW_MS
        if (detector.onWindow(levelDb, analyzeWindow(chunk, SR), WINDOW_MS, now)) fired = true
    }
    return fired
}

class CryCalibrationTest {
    // --- ALRM-2: the single sensitivity dial ------------------------------------------------

    @Test
    fun `ALRM-2 sensitivity runs 1 to 10 and defaults to the middle`() {
        assertEquals(1, Settings.SENSITIVITY_MIN)
        assertEquals(10, Settings.SENSITIVITY_MAX)
        assertEquals(5, Settings.SENSITIVITY_DEFAULT)
        assertEquals(Settings.SENSITIVITY_DEFAULT, Settings().alarmSensitivity)
    }

    @Test
    fun `ALRM-2 higher sensitivity means quieter sound can trigger`() {
        for (s in Settings.SENSITIVITY_MIN until Settings.SENSITIVITY_MAX) {
            assertTrue(
                "sensitivity ${s + 1} must demand less loudness than $s",
                sensitivityThresholdDb(s + 1) < sensitivityThresholdDb(s),
            )
        }
    }

    @Test
    fun `ALRM-2 a quiet cry alarms at high sensitivity but not at low`() {
        val quietCryDb = 6.0
        assertTrue(cryFires(sensitivityThresholdDb(Settings.SENSITIVITY_MAX), quietCryDb))
        assertFalse(cryFires(sensitivityThresholdDb(Settings.SENSITIVITY_DEFAULT), quietCryDb))
        assertFalse(cryFires(sensitivityThresholdDb(Settings.SENSITIVITY_MIN), quietCryDb))
    }

    @Test
    fun `ALRM-2 a sensitivity outside the scale is treated as its nearest end`() {
        // Stored settings can be old or hand-edited; the mapping must never explode or go wild.
        assertEquals(sensitivityThresholdDb(Settings.SENSITIVITY_MIN), sensitivityThresholdDb(-3), 1e-9)
        assertEquals(sensitivityThresholdDb(Settings.SENSITIVITY_MAX), sensitivityThresholdDb(99), 1e-9)
    }

    // --- ALRM-15: which alarms ask for an answer ---------------------------------------------

    @Test
    fun `ALRM-15 only the crying alarm asks whether it was a false alarm`() {
        assertTrue(asksForCryFeedback(AlarmKind.BABY_NOISE))
        assertFalse(asksForCryFeedback(AlarmKind.FEED_DOWN))
    }

    @Test
    fun `ALRM-15 the answer goes to the camera that alarmed, once, and only while one is pending`() {
        val applied = mutableListOf<Pair<String, Boolean>>()
        MonitorHub.onCryFeedback = { did, wasCry -> applied += did to wasCry }
        try {
            MonitorHub.submitCryFeedback(true) // nothing pending: a stray tap learns nothing
            assertTrue(applied.isEmpty())

            MonitorHub.pendingCryFeedback.value = "cam-1"
            MonitorHub.pendingCryFeedback.value = "cam-2" // a newer alarm replaces the question
            MonitorHub.submitCryFeedback(false)
            assertEquals(listOf("cam-2" to false), applied)
            assertEquals(null, MonitorHub.pendingCryFeedback.value) // asked once, answered once

            MonitorHub.pendingCryFeedback.value = "cam-1"
            MonitorHub.dismissCryFeedback() // optional: dismissing learns nothing (ALRM-15)
            assertEquals(null, MonitorHub.pendingCryFeedback.value)
            assertEquals(1, applied.size)
        } finally {
            MonitorHub.onCryFeedback = null
            MonitorHub.pendingCryFeedback.value = null
        }
    }

    // --- ALRM-16: bounded learning ------------------------------------------------------------

    @Test
    fun `ALRM-16 a false-alarm answer makes triggering one step harder`() {
        assertEquals(1, afterFalseAlarm(0))
        val before = effectiveThresholdDb(Settings.SENSITIVITY_DEFAULT, 0)
        val after = effectiveThresholdDb(Settings.SENSITIVITY_DEFAULT, afterFalseAlarm(0))
        assertTrue("a false alarm must raise the bar", after > before)

        // Integration: a cry just over the slider's bar fired before the answer, not after.
        val borderline = before + 1.0
        assertTrue(cryFires(before, borderline))
        assertFalse(cryFires(after, borderline))
    }

    @Test
    fun `ALRM-16 a confirmed cry undoes one step`() {
        assertEquals(0, afterRealCry(afterFalseAlarm(0)))
        assertEquals(1, afterRealCry(2))
    }

    @Test
    fun `ALRM-16 learning is bounded and never goes below the slider setting`() {
        // However many false alarms are reported, the steps stop at the cap...
        var steps = 0
        repeat(20) { steps = afterFalseAlarm(steps) }
        assertEquals(CALIBRATION_MAX_STEPS, steps)
        // ...and however many cries are confirmed, tuning never drops below the slider's own bar.
        repeat(20) { steps = afterRealCry(steps) }
        assertEquals(0, steps)
        assertEquals(
            sensitivityThresholdDb(Settings.SENSITIVITY_DEFAULT),
            effectiveThresholdDb(Settings.SENSITIVITY_DEFAULT, 0),
            1e-9,
        )
    }

    @Test
    fun `ALRM-16 corrupt stored steps are treated as their nearest bound`() {
        assertEquals(
            effectiveThresholdDb(Settings.SENSITIVITY_DEFAULT, 0),
            effectiveThresholdDb(Settings.SENSITIVITY_DEFAULT, -7),
            1e-9,
        )
        assertEquals(
            effectiveThresholdDb(Settings.SENSITIVITY_DEFAULT, CALIBRATION_MAX_STEPS),
            effectiveThresholdDb(Settings.SENSITIVITY_DEFAULT, 99),
            1e-9,
        )
    }

    @Test
    fun `ALRM-16 fully tuned-down detection still hears a loud cry — it never learns itself deaf`() {
        val maxTuned = effectiveThresholdDb(Settings.SENSITIVITY_DEFAULT, CALIBRATION_MAX_STEPS)
        assertTrue(cryFires(maxTuned, levelDb = maxTuned + 2.0))
        // And the trigger point stays on the level scale, where it can be seen and tuned.
        for (s in Settings.SENSITIVITY_MIN..Settings.SENSITIVITY_MAX) {
            assertTrue(effectiveThresholdDb(s, CALIBRATION_MAX_STEPS) <= LevelMeter.LEVEL_MAX)
        }
    }
}

File: windows/tests/BabyMonitor.Core.Tests/CryCalibrationTest.cs
using BabyMonitor.Core.Data;
using BabyMonitor.Core.Dsp;
using BabyMonitor.Core.Monitor;
using Xunit;

namespace BabyMonitor.Core.Tests;

// ALRM-2 (one sensitivity dial) and ALRM-15/16/17 (learning from the parent's answers). The mapping
// and the learning are pure logic; detector integration runs the REAL analysis.

public class CryCalibrationTest
{
    private const int Window = 2048;
    private const long WindowMs = Window * 1000L / Sounds.Sr;

    /// <summary>Feed a real synthesised cry to a detector at this bar; did it ever alarm?</summary>
    private static bool CryFires(double thresholdDb, double levelDb, double seconds = 4.0)
    {
        var detector = new BabyNoiseDetector { Enabled = true, ThresholdDb = thresholdDb };
        long now = 0;
        var fired = false;
        var windows = (int)(seconds * Sounds.Sr / Window);
        var pcm = Sounds.BabyCry(Window * windows);
        for (var w = 0; w < windows; w++)
        {
            now += WindowMs;
            var metrics = Analysis.AnalyzeWindow(pcm[(w * Window)..((w + 1) * Window)], Sounds.Sr);
            if (detector.OnWindow(levelDb, metrics, WindowMs, now))
            {
                fired = true;
            }
        }

        return fired;
    }

    // --- ALRM-2: the single sensitivity dial ------------------------------------------------

    [Fact(DisplayName = "ALRM-2 sensitivity runs 1 to 10 and defaults to the middle")]
    public void SensitivityScale()
    {
        Assert.Equal(1, Settings.SensitivityMin);
        Assert.Equal(10, Settings.SensitivityMax);
        Assert.Equal(5, Settings.SensitivityDefault);
        Assert.Equal(Settings.SensitivityDefault, new Settings().AlarmSensitivity);
    }

    [Fact(DisplayName = "ALRM-2 higher sensitivity means quieter sound can trigger")]
    public void HigherSensitivityNeedsLessLoudness()
    {
        for (var s = Settings.SensitivityMin; s < Settings.SensitivityMax; s++)
        {
            Assert.True(
                CryCalibration.SensitivityThresholdDb(s + 1) < CryCalibration.SensitivityThresholdDb(s),
                $"sensitivity {s + 1} must demand less loudness than {s}");
        }
    }

    [Fact(DisplayName = "ALRM-2 a quiet cry alarms at high sensitivity but not at low")]
    public void QuietCryOnlyAtHighSensitivity()
    {
        const double quietCryDb = 6.0;
        Assert.True(CryFires(CryCalibration.SensitivityThresholdDb(Settings.SensitivityMax), quietCryDb));
        Assert.False(CryFires(CryCalibration.SensitivityThresholdDb(Settings.SensitivityDefault), quietCryDb));
        Assert.False(CryFires(CryCalibration.SensitivityThresholdDb(Settings.SensitivityMin), quietCryDb));
    }

    [Fact(DisplayName = "ALRM-2 a sensitivity outside the scale is treated as its nearest end")]
    public void SensitivityIsClamped()
    {
        // Stored settings can be old or hand-edited; the mapping must never explode or go wild.
        Assert.Equal(
            CryCalibration.SensitivityThresholdDb(Settings.SensitivityMin),
            CryCalibration.SensitivityThresholdDb(-3),
            9);
        Assert.Equal(
            CryCalibration.SensitivityThresholdDb(Settings.SensitivityMax),
            CryCalibration.SensitivityThresholdDb(99),
            9);
    }

    // --- ALRM-15: which alarms ask for an answer ---------------------------------------------

    [Fact(DisplayName = "ALRM-15 only the crying alarm asks whether it was a false alarm")]
    public void OnlyCryingAsks()
    {
        Assert.True(CryCalibration.AsksForCryFeedback(AlarmKind.BabyNoise));
        Assert.False(CryCalibration.AsksForCryFeedback(AlarmKind.FeedDown));
    }

    [Fact(DisplayName = "ALRM-15 the answer goes to the camera that alarmed — once — and only while one is pending")]
    public void TheAnswerGoesToTheRightCamera()
    {
        var applied = new List<(string Did, bool WasCry)>();
        MonitorHub.OnCryFeedback = (did, wasCry) => applied.Add((did, wasCry));
        try
        {
            MonitorHub.SubmitCryFeedback(true); // nothing pending: a stray click learns nothing
            Assert.Empty(applied);

            MonitorHub.PendingCryFeedback.Value = "cam-1";
            MonitorHub.PendingCryFeedback.Value = "cam-2"; // a newer alarm replaces the question
            MonitorHub.SubmitCryFeedback(false);
            Assert.Equal(new[] { ("cam-2", false) }, applied);
            Assert.Null(MonitorHub.PendingCryFeedback.Value); // asked once, answered once

            MonitorHub.PendingCryFeedback.Value = "cam-1";
            MonitorHub.DismissCryFeedback(); // optional: dismissing learns nothing (ALRM-15)
            Assert.Null(MonitorHub.PendingCryFeedback.Value);
            Assert.Single(applied);
        }
        finally
        {
            MonitorHub.OnCryFeedback = null;
            MonitorHub.PendingCryFeedback.Value = null;
        }
    }

    // --- ALRM-16: bounded learning ------------------------------------------------------------

    [Fact(DisplayName = "ALRM-16 a false-alarm answer makes triggering one step harder")]
    public void FalseAlarmRaisesTheBar()
    {
        Assert.Equal(1, CryCalibration.AfterFalseAlarm(0));
        var before = CryCalibration.EffectiveThresholdDb(Settings.SensitivityDefault, 0);
        var after = CryCalibration.EffectiveThresholdDb(
            Settings.SensitivityDefault,
            CryCalibration.AfterFalseAlarm(0));
        Assert.True(after > before, "a false alarm must raise the bar");

        // Integration: a cry just over the slider's bar fired before the answer, not after.
        var borderline = before + 1.0;
        Assert.True(CryFires(before, borderline));
        Assert.False(CryFires(after, borderline));
    }

    [Fact(DisplayName = "ALRM-16 a confirmed cry undoes one step")]
    public void ConfirmedCryUndoesAStep()
    {
        Assert.Equal(0, CryCalibration.AfterRealCry(CryCalibration.AfterFalseAlarm(0)));
        Assert.Equal(1, CryCalibration.AfterRealCry(2));
    }

    [Fact(DisplayName = "ALRM-16 learning is bounded and never goes below the slider setting")]
    public void LearningIsBounded()
    {
        // However many false alarms are reported, the steps stop at the cap...
        var steps = 0;
        for (var i = 0; i < 20; i++)
        {
            steps = CryCalibration.AfterFalseAlarm(steps);
        }

        Assert.Equal(CryCalibration.MaxSteps, steps);

        // ...and however many cries are confirmed, tuning never drops below the slider's own bar.
        for (var i = 0; i < 20; i++)
        {
            steps = CryCalibration.AfterRealCry(steps);
        }

        Assert.Equal(0, steps);
        Assert.Equal(
            CryCalibration.SensitivityThresholdDb(Settings.SensitivityDefault),
            CryCalibration.EffectiveThresholdDb(Settings.SensitivityDefault, 0),
            9);
    }

    [Fact(DisplayName = "ALRM-16 corrupt stored steps are treated as their nearest bound")]
    public void CorruptStepsAreClamped()
    {
        Assert.Equal(
            CryCalibration.EffectiveThresholdDb(Settings.SensitivityDefault, 0),
            CryCalibration.EffectiveThresholdDb(Settings.SensitivityDefault, -7),
            9);
        Assert.Equal(
            CryCalibration.EffectiveThresholdDb(Settings.SensitivityDefault, CryCalibration.MaxSteps),
            CryCalibration.EffectiveThresholdDb(Settings.SensitivityDefault, 99),
            9);
    }

    [Fact(DisplayName = "ALRM-16 fully tuned-down detection still hears a loud cry — it never learns itself deaf")]
    public void ItNeverLearnsItselfDeaf()
    {
        var maxTuned = CryCalibration.EffectiveThresholdDb(Settings.SensitivityDefault, CryCalibration.MaxSteps);
        Assert.True(CryFires(maxTuned, levelDb: maxTuned + 2.0));

        // And the trigger point stays on the level scale, where it can be seen and tuned.
        for (var s = Settings.SensitivityMin; s <= Settings.SensitivityMax; s++)
        {
            Assert.True(CryCalibration.EffectiveThresholdDb(s, CryCalibration.MaxSteps) <= LevelMeter.LevelMax);
        }
    }
}
